# Homework 3

## FINM 37500: Fixed Income Derivatives

### Mark Hendricks

#### Winter 2024

# 1. Modeling the Volatility Smile

## Swaption Vol Data

The file `data/swaption_vol_data_2024-02-20.xlsx` has market data on the implied volatility skews for swaptions. Note that it has several columns:
* `expry`: expiration of the swaption
* `tenor`: tenor of the underlying swap
* `model`: the model by which the volatility is quoted. (All are Black.)
* `-200`, `-100`, etc.: The strike listed as difference from ATM strike (bps). Note that ATM is considered to be the **forward swapa rate** which you can calculate.

Your data: ywill use a single row of this data for the `1x4` swaption.
* date: `2024-02-20`
* expiration: 1yr
* tenor: 4yrs

## Rate Data

The file `data/cap_quotes_2024-02-20.xlsx` gives 
* SOFR swap rates, 
* their associated discount factors
* their associated forward interest rates.

You will not need the cap data (flat or forward vols) for this problem.
* This cap data would be helpful in calibrating a binomial tree, but this problem focuses on Black's formula and SABR.

## The Swaption

Consider the following swaption with the following features:
* underlying is a fixed-for-floating (SOFR) swap
* the underlying swap has **quarterly** payment frequency
* this is a **payer** swaption, which gives the holder the option to **pay** the fixed swap rate and receive SOFR.

## 1.1
Calculate the (relevant) forward swap rate. That is, the one-year forward 4-year swap rate.

## 1.2
Price the swaptions at the quoted implied volatilites and corresponding strikes, all using the just-calculated forward swap rate as the underlying.

## 1.3
To consider how the expiration and tenor matter, calculate the prices of a few other swaptions for comparison. 
* No need to get other implied vol quotes--just use the ATM implied vol you have for the 1x2 above. (Here we are just interested in how Black's formula changes with changes in tenor and expiration.
* No need to calculate for all the strikes--just do the ATM strike.

Alternate swaptions
* The 3mo x 4yr swaption
* The 2yr x 4yr swaption
* the 1yr x 2yr swaption

Report these values and compare them to the price of the `1y x 4y` swaption.

In [1]:
import numpy as np
import pandas as pd

import platform
if platform.system() == "Windows":
    BASE_DIR = r'C:\Users\Alex\Desktop\Academic\UChicago\FINM 37500\finm-fiderivs-2024'
else:
    BASE_DIR = r'/Users/alexhuang/Documents/Academic/FINM 37500/finm-fiderivs-2024'


import sys, os
sys.path.append(BASE_DIR)

from cmds.binomial import *
from cmds.ratecurves import *
from cmds.volskew import *
from cmds.ficcvol import *
from cmds.fi_binomial_model import FIBinomialModel

In [2]:
vol = pd.read_excel('../data/swaption_vol_data.xlsx')
vol

reference instrument  model       date  expiration  tenor   -200   -100  \
0      SOFR   swaption  black 2024-02-20           1      4  54.54  40.37   

     -50    -25      0     25     50    100    200  
0  35.94  34.23  32.83  31.71  30.86  29.83  29.54

In [3]:
cap_curves = pd.read_excel('../data/cap_curves_2024-02-20.xlsx')
cap_curves.set_index('tenor', inplace=True)
cap_curves

swap rates  spot rates  discounts  forwards  flat vols  fwd vols
tenor                                                                  
0.25     0.052211    0.052211   0.987115       NaN        NaN       NaN
0.50     0.051540    0.051535   0.974722  0.050860   0.166025  0.166025
0.75     0.050506    0.050490   0.963069  0.048400   0.191290  0.210648
1.00     0.049284    0.049250   0.952230  0.045531   0.216554  0.254312
1.25     0.047631    0.047565   0.942608  0.040831   0.260043  0.361247
1.50     0.046235    0.046141   0.933499  0.039030   0.292615  0.380930
1.75     0.045059    0.044939   0.924774  0.037738   0.315878  0.388953
2.00     0.044133    0.043994   0.916212  0.037382   0.331443  0.386643
2.25     0.043173    0.043011   0.908230  0.035151   0.340919  0.376247
2.50     0.042461    0.042283   0.900188  0.035738   0.345916  0.363764
2.75     0.041934    0.041745   0.892076  0.036373   0.348042  0.353513
3.00     0.041542    0.041347   0.883906  0.036973   0.348907  0.349207
3.25     0.041101    0.040896   0.876134  0.035482   0.349778  0.350339
3.50     0.040690    0.040473   0.868539  0.034978   0.350549  0.351556
3.75     0.040416    0.040195   0.860726  0.036309   0.350772  0.349233
4.00     0.040197    0.039974   0.852910  0.036657   0.349998  0.339501
4.25     0.039940    0.039710   0.845410  0.035486   0.347970  0.322525
4.50     0.039699    0.039461   0.838026  0.035245   0.345200  0.309744
4.75     0.039540    0.039301   0.830464  0.036421   0.342387  0.306238
5.00     0.039432    0.039195   0.822816  0.037180   0.340234  0.311414
5.25     0.039329    0.039093   0.815264  0.037055   0.339208  0.324078
5.50     0.039225    0.038990   0.807828  0.036819   0.338844  0.332488
5.75     0.039127    0.038891   0.800481  0.036713   0.338442  0.331264
6.00     0.039038    0.038802   0.793191  0.036762   0.337304  0.317965
6.25     0.038964    0.038728   0.785930  0.036954   0.334939  0.294203
6.50     0.038902    0.038668   0.778698  0.037151   0.331698  0.273998
6.75     0.038849    0.038617   0.771503  0.037305   0.328142  0.262325
7.00     0.038804    0.038574   0.764355  0.037404   0.324831  0.260895
7.25     0.038763    0.038535   0.757265  0.037451   0.322182  0.268693
7.50     0.038726    0.038501   0.750230  0.037507   0.320050  0.275091
7.75     0.038695    0.038471   0.743246  0.037589   0.318148  0.276397
8.00     0.038668    0.038447   0.736306  0.037699   0.316187  0.271592
8.25     0.038646    0.038429   0.729407  0.037833   0.313948  0.261350
8.50     0.038629    0.038415   0.722550  0.037964   0.311483  0.251699
8.75     0.038616    0.038406   0.715735  0.038085   0.308914  0.244536
9.00     0.038606    0.038400   0.708966  0.038190   0.306360  0.240251
9.25     0.038598    0.038396   0.702246  0.038278   0.303943  0.239254
9.50     0.038593    0.038395   0.695578  0.038344   0.301782  0.241978
9.75     0.038588    0.038395   0.688967  0.038385   0.300000  0.248893
10.00    0.038584    0.038395   0.682416  0.038398   0.298715  0.260524

## 1.1

In [4]:
start = 1
end = start + 4

forward_swap_rate = calc_fwdswaprate(cap_curves['discounts'], start, end, freqswap=4)
forward_swap_rate

0.03672212061985555

## 1.2

In [5]:
strikerange = np.array(vol.columns[-9:].tolist())
vols = vol[strikerange] / 100
strikes = forward_swap_rate + strikerange/100/100
# display(strikerange)
# display(vols)
# display(forward_swap_rate)
# display(strikes)

period_fwd = cap_curves.index.get_loc(start)
period_swap = cap_curves.index.get_loc(end) + 1
# display(period_fwd)
# display(period_swap)

discount = cap_curves['discounts'].iloc[period_fwd+1 : period_swap].sum() / 4
print(cap_curves['discounts'].iloc[period_fwd+1 : period_swap])
# display(cap_curves.index[period_fwd+1])
# display(cap_curves.index[period_swap])
# display(discount)
# print(start, vols, strikes, forward_swap_rate, discount, sep='\n')

quotes = vols.copy()
quotes.loc['price'] = 100 * blacks_formula(start, vols, strikes, forward_swap_rate, discount, isCall=True)[0]
quotes.loc['strike'] = strikes
quotes = quotes.rename(index={0: 'implied vol'})
quotes = quotes.loc[['strike', 'implied vol', 'price']]
quotes

tenor
1.25    0.942608
1.50    0.933499
1.75    0.924774
2.00    0.916212
2.25    0.908230
2.50    0.900188
2.75    0.892076
3.00    0.883906
3.25    0.876134
3.50    0.868539
3.75    0.860726
4.00    0.852910
4.25    0.845410
4.50    0.838026
4.75    0.830464
5.00    0.822816
Name: discounts, dtype: float64


-200      -100       -50       -25         0        25  \
strike       0.016722  0.026722  0.031722  0.034222  0.036722  0.039222   
implied vol  0.545400  0.403700  0.359400  0.342300  0.328300  0.317100   
price        7.203088  4.065315  2.739848  2.174850  1.687383  1.281199   

                   50       100       200  
strike       0.041722  0.046722  0.056722  
implied vol  0.308600  0.298300  0.295400  
price        0.955570  0.513281  0.147265

## 1.3

In [6]:
expiries = [.25,2,1]
tenors = [4,4,2]
fwdswaps = np.full(len(expiries), np.nan)

quotes_alt = pd.DataFrame(columns=['expiry', 'tenor', 'price'])

idstrikeATM = np.where(strikerange==0)[0][0]
# display(idstrikeATM)
strikeATM = strikes[idstrikeATM]
# display(strikeATM)
volATM = vols.iloc[0,idstrikeATM]
# display(volATM)

for i in range(len(fwdswaps)):
    print(expiries[i], expiries[i]+tenors[i], sep='\n')
    fwdswaps[i] = calc_fwdswaprate(cap_curves['discounts'], expiries[i], expiries[i]+tenors[i], freqswap=4)
    
    period0 = cap_curves.index.get_loc(expiries[i])
    period1 = cap_curves.index.get_loc(expiries[i]+tenors[i])+1
    
    # print(period0, period1)
    # print(expiries[i])
    discount = cap_curves['discounts'].iloc[period0+1 : period1].sum() / 4
    # print(cap_curves['discounts'].iloc[period0+1 : period1])
    # print(discount)
    # print(expiries[i], volATM, strikeATM, forward_swap_rate, discount, sep='\n')
    # print()
    quotes_alt.loc[i,['expiry', 'tenor']] = [expiries[i], tenors[i]]
    quotes_alt.loc[i, 'price'] = 100 * blacks_formula(expiries[i], volATM, strikeATM, forward_swap_rate, discount, isCall=True)
    
quotes_alt

0.25
4.25
2
6
1
3


expiry tenor     price
0   0.25     4  0.870475
1      2     4  2.291364
2      1     2  0.874004

***

# 2. Pricing w/ BDT

Use the data in `cap_curves_2024-02-20.xlsx`.

## 2.1

Calibrate the BDT Tree
* theta to fit the term structure discounts.
* sigma to fit the fwd vols from the cap data.

Report the rate tree through $T=5$. Report trees for rates compounded
* continuously
* annually

In [7]:
sig = cap_curves['fwd vols']
sig.iloc[0] = sig.iloc[1]
sig

tenor
0.25     0.166025
0.50     0.166025
0.75     0.210648
1.00     0.254312
1.25     0.361247
1.50     0.380930
1.75     0.388953
2.00     0.386643
2.25     0.376247
2.50     0.363764
2.75     0.353513
3.00     0.349207
3.25     0.350339
3.50     0.351556
3.75     0.349233
4.00     0.339501
4.25     0.322525
4.50     0.309744
4.75     0.306238
5.00     0.311414
5.25     0.324078
5.50     0.332488
5.75     0.331264
6.00     0.317965
6.25     0.294203
6.50     0.273998
6.75     0.262325
7.00     0.260895
7.25     0.268693
7.50     0.275091
7.75     0.276397
8.00     0.271592
8.25     0.261350
8.50     0.251699
8.75     0.244536
9.00     0.240251
9.25     0.239254
9.50     0.241978
9.75     0.248893
10.00    0.260524
Name: fwd vols, dtype: float64

In [8]:
disc = cap_curves['discounts']
theta_params, rate_tree = estimate_theta(sig, disc)

In [9]:
pd.DataFrame(theta_params).T

time   0.00      0.25      0.50      0.75      1.00      1.25      1.50  \
theta   NaN -0.117846 -0.262526 -0.359865 -0.720504 -0.459257 -0.384915   

time       1.75      2.00      2.25  ...      7.50      7.75      8.00  \
theta -0.244081 -0.396789 -0.033089  ...  0.125541  0.134603  0.153908   

time       8.25      8.50      8.75      9.00      9.25      9.50      9.75  
theta  0.169295  0.178789  0.182346  0.179249  0.168698  0.149782  0.121431  

[1 rows x 40 columns]

In [10]:
# Continuously
rate_tree = rate_tree[[col for col in rate_tree.columns if col <= 5]]
rate_tree = rate_tree.dropna(how='all')
rate_tree.round(3)

time    0.00   0.25   0.50   0.75   1.00   1.25   1.50   1.75   2.00   2.25  \
state                                                                         
0      0.052  0.055  0.057  0.059  0.059  0.064  0.070  0.080  0.088  0.104   
1        NaN  0.046  0.048  0.050  0.050  0.054  0.060  0.068  0.074  0.088   
2        NaN    NaN  0.039  0.041  0.041  0.044  0.048  0.055  0.060  0.072   
3        NaN    NaN    NaN  0.031  0.031  0.034  0.037  0.043  0.047  0.056   
4        NaN    NaN    NaN    NaN  0.022  0.024  0.026  0.030  0.033  0.039   
5        NaN    NaN    NaN    NaN    NaN  0.016  0.018  0.020  0.022  0.026   
6        NaN    NaN    NaN    NaN    NaN    NaN  0.012  0.014  0.015  0.018   
7        NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.009  0.010  0.012   
8        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.007  0.008   
9        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.006   
10       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
11       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
14       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
15       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
16       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
17       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
18       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
19       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
20       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

time   ...   2.75   3.00   3.25   3.50   3.75   4.00   4.25   4.50   4.75  \
state  ...                                                                  
0      ...  0.150  0.171  0.200  0.247  0.298  0.343  0.406  0.500  0.608   
1      ...  0.127  0.145  0.170  0.210  0.252  0.291  0.344  0.424  0.515   
2      ...  0.103  0.117  0.137  0.170  0.204  0.236  0.279  0.343  0.417   
3      ...  0.080  0.091  0.107  0.132  0.158  0.183  0.216  0.266  0.323   
4      ...  0.056  0.063  0.074  0.092  0.110  0.127  0.151  0.185  0.225   
5      ...  0.038  0.043  0.051  0.063  0.075  0.087  0.103  0.127  0.154   
6      ...  0.026  0.029  0.034  0.042  0.051  0.059  0.070  0.086  0.104   
7      ...  0.018  0.020  0.023  0.029  0.035  0.040  0.047  0.058  0.071   
8      ...  0.012  0.014  0.016  0.020  0.024  0.027  0.033  0.040  0.049   
9      ...  0.008  0.010  0.011  0.014  0.017  0.019  0.023  0.028  0.034   
10     ...  0.006  0.007  0.008  0.010  0.012  0.013  0.016  0.020  0.024   
11     ...  0.004  0.005  0.006  0.007  0.008  0.009  0.011  0.014  0.017   
12     ...    NaN  0.003  0.004  0.005  0.006  0.007  0.008  0.010  0.012   
13     ...    NaN    NaN  0.003  0.003  0.004  0.005  0.006  0.007  0.008   
14     ...    NaN    NaN    NaN  0.002  0.003  0.003  0.004  0.005  0.006   
15     ...    NaN    NaN    NaN    NaN  0.002  0.002  0.003  0.003  0.004   
16     ...    NaN    NaN    NaN    NaN    NaN  0.002  0.002  0.002  0.003   
17     ...    NaN    NaN    NaN    NaN    NaN    NaN  0.001  0.002  0.002   
18     ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.001  0.002   
19     ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.001   
20     ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

time    5.00  
state         
0      0.721  
1      0.611  
2      0.495  
3      0.384  
4      0.267  
5      0.183  
6      0.124  
7      0.084  
8      0.058  
9      0.040  
10     0.028  
11     0.020  
12     0.014  
13     0.010  
14     0.007  
15     0.005  
16     0.004  
17     0.003  
18     0.002  
19     0.001  
20     0.001  

[21 rows x 21 columns]

In [11]:
# Annually
rate_tree_annual = np.exp(rate_tree) - 1
rate_tree_annual.round(3)

time    0.00   0.25   0.50   0.75   1.00   1.25   1.50   1.75   2.00   2.25  \
state                                                                         
0      0.053  0.056  0.059  0.061  0.061  0.066  0.073  0.084  0.092  0.110   
1        NaN  0.047  0.049  0.051  0.051  0.056  0.061  0.070  0.077  0.092   
2        NaN    NaN  0.040  0.041  0.041  0.045  0.049  0.057  0.062  0.074   
3        NaN    NaN    NaN  0.032  0.032  0.035  0.038  0.044  0.048  0.057   
4        NaN    NaN    NaN    NaN  0.022  0.024  0.026  0.030  0.033  0.039   
5        NaN    NaN    NaN    NaN    NaN  0.016  0.018  0.021  0.022  0.027   
6        NaN    NaN    NaN    NaN    NaN    NaN  0.012  0.014  0.015  0.018   
7        NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.009  0.010  0.012   
8        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.007  0.008   
9        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.006   
10       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
11       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
13       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
14       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
15       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
16       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
17       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
18       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
19       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
20       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

time   ...   2.75   3.00   3.25   3.50   3.75   4.00   4.25   4.50   4.75  \
state  ...                                                                  
0      ...  0.162  0.186  0.222  0.281  0.347  0.410  0.501  0.649  0.837   
1      ...  0.136  0.156  0.185  0.233  0.287  0.337  0.411  0.527  0.673   
2      ...  0.109  0.124  0.147  0.185  0.226  0.266  0.321  0.409  0.518   
3      ...  0.083  0.095  0.112  0.141  0.172  0.200  0.241  0.305  0.382   
4      ...  0.057  0.065  0.077  0.096  0.117  0.136  0.163  0.204  0.253   
5      ...  0.039  0.044  0.052  0.065  0.078  0.091  0.108  0.135  0.166   
6      ...  0.026  0.030  0.035  0.043  0.052  0.061  0.072  0.090  0.110   
7      ...  0.018  0.020  0.024  0.029  0.035  0.041  0.049  0.060  0.073   
8      ...  0.012  0.014  0.016  0.020  0.024  0.028  0.033  0.041  0.050   
9      ...  0.008  0.010  0.011  0.014  0.017  0.019  0.023  0.028  0.034   
10     ...  0.006  0.007  0.008  0.010  0.012  0.014  0.016  0.020  0.024   
11     ...  0.004  0.005  0.006  0.007  0.008  0.010  0.011  0.014  0.017   
12     ...    NaN  0.003  0.004  0.005  0.006  0.007  0.008  0.010  0.012   
13     ...    NaN    NaN  0.003  0.003  0.004  0.005  0.006  0.007  0.008   
14     ...    NaN    NaN    NaN  0.002  0.003  0.003  0.004  0.005  0.006   
15     ...    NaN    NaN    NaN    NaN  0.002  0.002  0.003  0.003  0.004   
16     ...    NaN    NaN    NaN    NaN    NaN  0.002  0.002  0.002  0.003   
17     ...    NaN    NaN    NaN    NaN    NaN    NaN  0.001  0.002  0.002   
18     ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.001  0.002   
19     ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.001   
20     ...    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

time    5.00  
state         
0      1.057  
1      0.842  
2      0.640  
3      0.468  
4      0.307  
5      0.200  
6      0.132  
7      0.088  
8      0.059  
9      0.041  
10     0.029  
11     0.020  
12     0.014  
13     0.010  
14     0.007  
15     0.005  
16     0.004  
17     0.003  
18     0.002  
19     0.001  
20     0.001  

[21 rows x 21 columns]

## 2.2

Use a tree to price a vanilla fixed-rate, 5-year bond with coupon rate equal to the forward swap rate calculated in problem `1.1.`

In [12]:
FACE = 100
T = 5
compound = 4
cpn = forward_swap_rate
cpn_freq = 2

In [13]:
payoff = lambda r: payoff_bond(r, 1/compound, facevalue = FACE * (1 + cpn / cpn_freq))
cash_flow_tree = construct_bond_cftree(T, compound, cpn)
# cash_flow_tree
# rate_tree.loc[cash_flow_tree.index, cash_flow_tree.columns]
bond_tree = bintree_pricing(payoff=payoff, ratetree=rate_tree.loc[cash_flow_tree.index, cash_flow_tree.columns], cftree=cash_flow_tree)
# bond_tree.round(2)
cash_flow_tree

time   0.00  0.25      0.50  0.75      1.00  1.25      1.50  1.75      2.00  \
state                                                                         
0       0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
1       0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
2       0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
3       0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
4       0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
5       0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
6       0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
7       0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
8       0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
9       0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
10      0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
11      0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
12      0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
13      0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
14      0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
15      0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
16      0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
17      0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
18      0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   
19      0.0   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   

time   2.25      2.50  2.75      3.00  3.25      3.50  3.75      4.00  4.25  \
state                                                                         
0       0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
1       0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
2       0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
3       0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
4       0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
5       0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
6       0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
7       0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
8       0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
9       0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
10      0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
11      0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
12      0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
13      0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
14      0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
15      0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
16      0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
17      0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
18      0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   
19      0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0  1.836106   0.0   

time       4.50  4.75  
state                  
0      1.836106   0.0  
1      1.836106   0.0  
2      1.836106   0.0  
3      1.836106   0.0  
4      1.836106   0.0  
5      1.836106   0.0  
6      1.836106   0.0  
7      1.836106   0.0  
8      1.836106   0.0  
9      1.836106   0.0  
10     1.836106   0.0  
11     1.836106   0.0  
12     1.836106   0.0  
13     1.836106   0.0  
14     1.836106   0.0  
15     1.836106   0.0  
16     1.836106   0.0  
17     1.836106   0.0  
18     1.836106   0.0  
19     1.836106   

In [14]:
bond_tree[0][0].round(2)

98.7

## 2.3

We will calculate the binomial tree for the 5-year swap, but here we do so by valuing the swap as...

$$\text{payer swap} = \text{floating rate note} - \text{fixed-rate bond}$$

Recall for the Floating-Rate Note:
* It has par value of 100 at each reset date.
* Every node is a reset date given the assumptions of the swap timing.

Report the tree for the 5-year swap.

In [15]:
swap_tree = 100 - bond_tree
swap_tree

time       0.00      0.25      0.50      0.75       1.00       1.25  \
state                                                                 
0      1.298017  2.510201  5.651032  7.100999  10.514763  12.302403   
1           NaN -2.490852  0.355579  1.495591   4.593869   6.061836   
2           NaN       NaN -4.054385 -3.202016  -0.411521   0.721689   
3           NaN       NaN       NaN -6.949468  -4.423434  -3.592023   
4           NaN       NaN       NaN       NaN  -7.491460  -6.903965   
5           NaN       NaN       NaN       NaN        NaN  -9.260424   
6           NaN       NaN       NaN       NaN        NaN        NaN   
7           NaN       NaN       NaN       NaN        NaN        NaN   
8           NaN       NaN       NaN       NaN        NaN        NaN   
9           NaN       NaN       NaN       NaN        NaN        NaN   
10          NaN       NaN       NaN       NaN        NaN        NaN   
11          NaN       NaN       NaN       NaN        NaN        NaN   
12          NaN       NaN       NaN       NaN        NaN        NaN   
13          NaN       NaN       NaN       NaN        NaN        NaN   
14          NaN       NaN       NaN       NaN        NaN        NaN   
15          NaN       NaN       NaN       NaN        NaN        NaN   
16          NaN       NaN       NaN       NaN        NaN        NaN   
17          NaN       NaN       NaN       NaN        NaN        NaN   
18          NaN       NaN       NaN       NaN        NaN        NaN   
19          NaN       NaN       NaN       NaN        NaN        NaN   

time        1.50       1.75       2.00       2.25       2.50       2.75  \
state                                                                     
0      15.998539  17.910126  21.588536  23.430670  26.852265  28.119317   
1       9.459525  11.105595  14.574140  16.266768  19.630740  20.939393   
2       3.781849   5.095263   8.259762   9.675920  12.830607  14.008841   
3      -0.850388   0.131774   2.969946   4.059876   6.928394   7.872315   
4      -4.426549  -3.729171  -1.179424  -0.399229   2.179681   2.862305   
5      -6.976793  -6.490424  -4.156797  -3.612797  -1.258875  -0.783720   
6      -8.756178  -8.418601  -6.237811  -5.862027  -3.669406  -3.344127   
7            NaN  -9.751617  -7.675925  -7.415643  -5.333510  -5.110507   
8            NaN        NaN  -8.669714  -8.488107  -6.480672  -6.326032   
9            NaN        NaN        NaN  -9.233505  -7.277291  -7.168982   
10           NaN        NaN        NaN        NaN  -7.835101  -7.759162   
11           NaN        NaN        NaN        NaN        NaN  -8.174659   
12           NaN        NaN        NaN        NaN        NaN        NaN   
13           NaN        NaN        NaN        NaN        NaN        NaN   
14           NaN        NaN        NaN        NaN        NaN        NaN   
15           NaN        NaN        NaN        NaN        NaN        NaN   
16           NaN        NaN        NaN        NaN        NaN        NaN   
17           NaN        NaN        NaN        NaN        NaN        NaN   
18           NaN        NaN        NaN        NaN        NaN        NaN   
19           NaN        NaN        NaN        NaN        NaN        NaN   

time        3.00       3.25       3.50       3.75       4.00       4.25  \
state                                                                     
0      30.714301  31.064925  32.541225  31.157977  30.270626  26.228940   
1      23.694948  24.312833  26.182436  25.317292  25.086261  21.815863   
2      16.745375  17.451150  19.557953  19.107870  19.507627  17.060254   
3      10.455301  11.084371  13.246466  13.034271  13.909433  12.191186   
4       5.242770   5.707034   7.793617   7.654452   8.810023   7.583572   
5       1.430159   1.751623   3.759422   3.655649   5.007195   4.140491   
6      -1.252011  -1.037154   0.908178   0.824187   2.311537   1.698811   
7      -3.100953  -2.958217  -1.054689  -1.123230   0.459958   0.024436   
8      -4.370632  -4.274099  -2.3954

## 2.4



Report the binomial tree for the one-year swaption on a 4-year swap with **european** exercise.
* At expiration, the swap tree from 2.3 will have 4 years left, as desired for pricing the 1y-4y swaption.

In [16]:
payoff_swaption = lambda p: np.maximum(p, 0)

T = 1
tsteps = int(T * compound) + 1

swaption_tree = bintree_pricing(payoff_swaption, rate_tree.iloc[:tsteps,:tsteps], undertree=swap_tree.iloc[:tsteps,:tsteps])
swaption_tree.round(3)

time    0.00   0.25   0.50   0.75    1.00
state                                    
0      1.711  2.914  4.787  7.444  10.515
1        NaN  0.554  1.121  2.268   4.594
2        NaN    NaN  0.000  0.000   0.000
3        NaN    NaN    NaN  0.000   0.000
4        NaN    NaN    NaN    NaN   0.000

## 2.5

Compare the pricing of the 1y4y swaption from Black's formula in Section 1 vs the binomial tree.

In section 1, our price was roughly 1.69. Here, our price is about 1.71; they are close values.

## 2.6

Reprice the swaption using the BDT tree, but this time assuming it is **american**-style exercise.

In [17]:
bintree_pricing(payoff_swaption, rate_tree.iloc[:tsteps,:tsteps], undertree=swap_tree.iloc[:tsteps,:tsteps], style='american').round(3)

time    0.00   0.25   0.50   0.75    1.00
state                                    
0      1.922  3.340  5.651  7.444  10.515
1        NaN  0.554  1.121  2.268   4.594
2        NaN    NaN  0.000  0.000   0.000
3        NaN    NaN    NaN  0.000   0.000
4        NaN    NaN    NaN    NaN   0.000

***

# 3. Midcurve Swaptions

## 3.1 

Use the BDT tree from section 2 to price a **european** midcurve swaption 1y $\rightarrow$ 2y $\rightarrow$ 2y.

In [26]:
FACE = 100
T = 5
compound = 4
cpn = forward_swap_rate
cpn_freq = 4

In [27]:
payoff = lambda r: payoff_bond(r, 1/compound, facevalue = FACE * (1 + cpn / cpn_freq))
cash_flow_tree = construct_bond_cftree(T, compound, cpn, cpn_freq=cpn_freq)
cash_flow_tree.iloc[:, :12] = 0
delayed_swap_tree = bintree_pricing(payoff=payoff, ratetree=rate_tree.iloc[:int(T * compound),:int(T * compound)], cftree=cash_flow_tree)
delayed_swap_tree.iloc[:, 12:] = 100 - delayed_swap_tree.iloc[:, 12:]
delayed_swap_tree.iloc[:, :12] = 0
delayed_swap_tree

time   0.00  0.25  0.50  0.75  1.00  1.25  1.50  1.75  2.00  2.25  2.50  2.75  \
state                                                                           
0       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
2       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
5       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
6       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
7       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
8       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
9       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
10      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
11      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
12      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
13      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
14      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
15      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
16      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
17      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
18      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
19      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

time        3.00       3.25       3.50       3.75       4.00       4.25  \
state                                                                     
0      30.546699  31.791240  32.372377  31.880900  30.119641  26.969433   
1      23.558940  25.072782  26.042388  26.071234  24.958627  22.582447   
2      16.640276  18.245265  19.447635  19.894874  19.405095  17.855262   
3      10.378022  11.909910  13.164414  13.853174  13.832280  13.014783   
4       5.187963   6.559486   7.735214   8.502063   8.754990   8.434566   
5       1.391815   2.623912   3.718567   4.524629   4.968743   5.011977   
6      -1.278737  -0.150858   0.879769   1.708313   2.284896   2.584839   
7      -3.119637  -2.062255  -1.074490  -0.228675   0.441471   0.920442   
8      -4.383763  -3.371510  -2.409324  -1.547518  -0.809194  -0.204308   
9      -5.258799  -4.275458  -3.327923  -2.451609  -1.662697  -0.967926   
10     -5.871146  -4.907286  -3.968705  -3.080513  -2.254223  -1.494613   
11     -6.302645  -5.352811  -4.420630  -3.523880  -2.670741  -1.864646   
12     -6.606403  -5.666923  -4.739796  -3.837492  -2.965730  -2.126911   
13           NaN  -5.886750  -4.963444  -4.057634  -3.173226  -2.311796   
14           NaN        NaN  -5.118818  -4.210547  -3.317470  -2.440559   
15           NaN        NaN        NaN  -4.316258  -3.416895  -2.529156   
16           NaN        NaN        NaN        NaN  -3.485988  -2.590320   
17           NaN        NaN        NaN        NaN        NaN  -2.633816   
18           NaN        NaN        NaN        NaN        NaN        NaN   
19           NaN        NaN        NaN        NaN        NaN        NaN   

time        4.50       4.75  
state                        
0      21.787994  13.310502  
1      18.373034  11.271489  
2      14.718285   9.075285  
3      10.971198   6.921123  
4       7.234710   4.610935  
5       4.443365   2.893086  
6       2.465871   1.680991  
7       1.112958   0.854939  
8       0.202838   0.302365  
9      -0.411500  -0.068223  
10     -0.832706  -0.320537  
11     -1.127587  -0.496288  
12     -1.336597  -0.620757  
13     -1.484226  -0.708787  
14     -1.587314  -0.770427  
15     -1.658231  -0.812903  
16     -1.706882  -0.84188

In [20]:
payoff = lambda p: p

delayed_swap_tree = bintree_pricing(payoff=payoff, ratetree=rate_tree.iloc[:13,:13], undertree=delayed_swap_tree.iloc[:13,:13])
delayed_swap_tree

time   0.00  0.25  0.50  0.75  1.00  1.25  1.50  1.75  2.00  2.25  2.50  2.75  \
state                                                                           
0       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1       NaN   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
2       NaN   NaN   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3       NaN   NaN   NaN   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4       NaN   NaN   NaN   NaN   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
5       NaN   NaN   NaN   NaN   NaN   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
6       NaN   NaN   NaN   NaN   NaN   NaN   0.0   0.0   0.0   0.0   0.0   0.0   
7       NaN   NaN   NaN   NaN   NaN   NaN   NaN   0.0   0.0   0.0   0.0   0.0   
8       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   0.0   0.0   0.0   0.0   
9       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   0.0   0.0   0.0   
10      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   0.0   0.0   
11      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   0.0   
12      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

time   3.00  
state        
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  
5       0.0  
6       0.0  
7       0.0  
8       0.0  
9       0.0  
10      0.0  
11      0.0  
12      0.0

In [21]:
payoff_swaption = lambda p: np.maximum(p, 0)

T = 1
tsteps = int(T * compound) + 1

midcurve_swaption_tree = bintree_pricing(payoff_swaption, rate_tree.iloc[:tsteps,:tsteps], undertree=delayed_swap_tree.iloc[:tsteps,:tsteps])
midcurve_swaption_tree.round(3)

time   0.00  0.25  0.50  0.75  1.00
state                              
0       0.0   0.0   0.0   0.0   0.0
1       NaN   0.0   0.0   0.0   0.0
2       NaN   NaN   0.0   0.0   0.0
3       NaN   NaN   NaN   0.0   0.0
4       NaN   NaN   NaN   NaN   0.0

## 3.2

Price the **american** midcurve swaption 1y $\rightarrow$ 2y $\rightarrow$ 2y.

***

In [22]:
bintree_pricing(payoff_swaption, rate_tree.iloc[:tsteps,:tsteps], undertree=delayed_swap_tree.iloc[:tsteps,:tsteps], style='american').round(3)

time   0.00  0.25  0.50  0.75  1.00
state                              
0       0.0   0.0   0.0   0.0   0.0
1       NaN   0.0   0.0   0.0   0.0
2       NaN   NaN   0.0   0.0   0.0
3       NaN   NaN   NaN   0.0   0.0
4       NaN   NaN   NaN   NaN   0.0